src/main/java/org/terasology/scenario/internal/utilities/ArgumentParser.java

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright 2017 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.scenario.internal.utilities;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.assets.management.AssetManager;
+import org.terasology.entitySystem.Component;
+import org.terasology.entitySystem.entity.EntityManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.prefab.Prefab;
+import org.terasology.rendering.FontColor;
+import org.terasology.rendering.nui.Color;
+import org.terasology.rendering.nui.UIWidget;
+import org.terasology.rendering.nui.widgets.UIButton;
+import org.terasology.rendering.nui.widgets.UILabel;
+import org.terasology.scenario.components.actions.ArgumentContainerComponent;
+import org.terasology.scenario.components.actions.TextComponent;
+import org.terasology.scenario.components.information.BlockComponent;
+import org.terasology.scenario.components.information.IndentificationComponents.ScenarioBlockComponent;
+import org.terasology.scenario.components.information.IndentificationComponents.ScenarioPlayerEntityComponent;
+import org.terasology.scenario.components.information.InformationEnums;
+import org.terasology.scenario.components.information.PlayerComponent;
+import org.terasology.scenario.internal.events.evaluationEvents.EvaluateBlockDisplayEvent;
+import org.terasology.scenario.internal.events.evaluationEvents.EvaluateIntDisplayEvent;
+import org.terasology.scenario.internal.events.evaluationEvents.EvaluatePlayerDisplayEvent;
+import org.terasology.scenario.internal.events.evaluationEvents.EvaluateStringDisplayEvent;
+import org.terasology.scenario.internal.ui.EditLogicScreen;
+import org.terasology.scenario.internal.ui.EditParameterScreen;
+import org.terasology.world.block.BlockManager;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class ArgumentParser {
+    private Logger logger = LoggerFactory.getLogger(ArgumentParser.class);
+
+
+    private BlockManager blockManager;
+
+
+    private EntityManager entityManager;
+
+    private AssetManager assetManager;
+
+    private static ArgumentParser parser;
+
+    private List<String> keys;
+
+    private ArgumentParser() {
+
+    }
+
+    public static ArgumentParser getInstance() {

```

Comment:
Singletons are anti patterns. Storing variable data in static fields is bad for testing, limits the usage to 1 thread and might even result to bugs when the cleanup is not done properly or when it gets used after for example the used entity manager got disposed.